// @json-render/mcp/dist/app.mjs
import {
  buildAppHtml
} from "./chunk-UZ7LXWZJ.mjs";

// src/use-json-render-app.ts
import { useState, useEffect, useCallback, useRef } from "react";
import { App } from "@modelcontextprotocol/ext-apps";
function parseSpecFromToolResult(result) {
  const textContent = result.content?.find(
    (c) => c.type === "text"
  );
  if (!textContent?.text) return null;
  try {
    const parsed = JSON.parse(textContent.text);
    if (parsed && typeof parsed === "object" && "spec" in parsed) {
      return parsed.spec;
    }
    return parsed;
  } catch {
    return null;
  }
}
function useJsonRenderApp(options = {}) {
  const { name = "json-render", version = "1.0.0" } = options;
  const [spec, setSpec] = useState(null);
  const [loading, setLoading] = useState(true);
  const [connected, setConnected] = useState(false);
  const [error, setError] = useState(null);
  const appRef = useRef(null);
  useEffect(() => {
    const app = new App({ name, version });
    appRef.current = app;
    app.ontoolresult = (result) => {
      const parsed = parseSpecFromToolResult(result);
      if (parsed) {
        setSpec(parsed);
        setLoading(false);
      }
    };
    app.connect().then(() => {
      setConnected(true);
    }).catch((err) => {
      setError(err instanceof Error ? err : new Error(String(err)));
    });
    return () => {
      app.close().catch(() => {
      });
    };
  }, []);
  const callServerTool = useCallback(
    async (toolName, args = {}) => {
      if (!appRef.current) return;
      setLoading(true);
      try {
        const result = await appRef.current.callServerTool({
          name: toolName,
          arguments: args
        });
        const parsed = parseSpecFromToolResult(result);
        if (parsed) setSpec(parsed);
      } finally {
        setLoading(false);
      }
    },
    []
  );
  return {
    spec,
    connecting: !connected && !error,
    connected,
    error,
    loading,
    app: appRef.current,
    callServerTool
  };
}
export {
  buildAppHtml,
  useJsonRenderApp
};
